### PR TITLE
Remove deprecated numpy type aliases

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## What changes are proposed in this pull request?

- Closes #4045
- Removes deprecated numpy type aliases such as `np.int`

https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations

## Quick check that deprecated aliases are identical to python primitive types

```python
>>> import numpy as np
>>> np.str == str
True
>>> np.int == int
True
>>> np.long == int
True
>>> np.bool == bool
True
>>> np.object == object
True
>>> np.float == float
True
```

## Motivation

- Prevent MLflow from raising deprecated warnings
- Clean up pytest output logs:

image

## How is this patch tested?

Existing tests

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/sklearn/test_sklearn_autolog.py [hunk 1]
```diff
@@ -1874,10 +1874,10 @@ def test_is_metrics_value_loggable():
     is_metric_value_loggable = mlflow.sklearn._AutologgingMetricsManager.is_metric_value_loggable
     assert is_metric_value_loggable(3)
     assert is_metric_value_loggable(3.5)
-    assert is_metric_value_loggable(np.int(3))
+    assert is_metric_value_loggable(int(3))
     assert is_metric_value_loggable(np.float32(3.5))
     assert not is_metric_value_loggable(True)
-    assert not is_metric_value_loggable(np.bool(True))
+    assert not is_metric_value_loggable(bool(True))
```
Comment (harupy) [inline tests/sklearn/test_sklearn_autolog.py:1880]:
```suggestion
```

The line above does the same thing.

File: tests/spark/autologging/ml/test_pyspark_ml_autologging.py [hunk 2]
```diff
@@ -753,10 +753,10 @@ def test_is_metrics_value_loggable():
     is_metric_value_loggable = mlflow.pyspark.ml._AutologgingMetricsManager.is_metric_value_loggable
     assert is_metric_value_loggable(3)
     assert is_metric_value_loggable(3.5)
-    assert is_metric_value_loggable(np.int(3))
+    assert is_metric_value_loggable(int(3))
     assert is_metric_value_loggable(np.float32(3.5))
     assert not is_metric_value_loggable(True)
-    assert not is_metric_value_loggable(np.bool(True))
+    assert not is_metric_value_loggable(bool(True))
```
Comment (harupy) [inline tests/spark/autologging/ml/test_pyspark_ml_autologging.py:759]:
```suggestion
```

File: tests/sklearn/test_sklearn_autolog.py [hunk 3]
```diff
@@ -1874,10 +1874,10 @@ def test_is_metrics_value_loggable():
     is_metric_value_loggable = mlflow.sklearn._AutologgingMetricsManager.is_metric_value_loggable
     assert is_metric_value_loggable(3)
     assert is_metric_value_loggable(3.5)
-    assert is_metric_value_loggable(np.int(3))
+    assert is_metric_value_loggable(int(3))
```
Comment (harupy) [inline tests/sklearn/test_sklearn_autolog.py:1877]:
```suggestion
```

line 1875 does the same thing.

File: tests/spark/autologging/ml/test_pyspark_ml_autologging.py [hunk 4]
```diff
@@ -753,10 +753,10 @@ def test_is_metrics_value_loggable():
     is_metric_value_loggable = mlflow.pyspark.ml._AutologgingMetricsManager.is_metric_value_loggable
     assert is_metric_value_loggable(3)
     assert is_metric_value_loggable(3.5)
-    assert is_metric_value_loggable(np.int(3))
+    assert is_metric_value_loggable(int(3))
```
Comment (harupy) [inline tests/spark/autologging/ml/test_pyspark_ml_autologging.py:756]:
```suggestion
```

File: pytest.ini [hunk 5]
```diff
@@ -1,2 +1,6 @@
 [pytest]
 addopts = --color=yes --durations=5 --showlocals --verbose
+# Turn warnings for deprecated numpy type aliases into errors
+filterwarnings =
+  error:^`np\.[a-z]+` is a deprecated alias for the builtin.+:DeprecationWarning:mlflow
+  error:^`np\.[a-z]+` is a deprecated alias for the builtin.+:DeprecationWarning:tests
```
Comment (harupy) [inline pytest.ini:6]: This works as a test that ensures we don't use deprecated types aliases.

File: mlflow/types/utils.py [hunk 6]
```diff
@@ -238,7 +237,7 @@ def __call__(self, x):
             return DataType.string
         else:
             raise MlflowException(
-                "Unable to map 'np.object' type to MLflow DataType. np.object can"
+                "Unable to map 'object' type to MLflow DataType. object can"
```
Comment (harupy) [inline mlflow/types/utils.py:240]: This might sound strange since everything is an object in Python.

File: pytest.ini [hunk 7]
```diff
@@ -1,2 +1,7 @@
 [pytest]
 addopts = --color=yes --durations=5 --showlocals --verbose
+# Turn warnings for deprecated numpy type aliases into errors
+# https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
+filterwarnings =
+  error:^`np\.[a-z]+` is a deprecated alias for the builtin.+:DeprecationWarning:mlflow
```
Comment (harupy) [inline pytest.ini:6]: The format here is:

```
error:<warning message pattern>:<warning type>:<where warning originates from>
```

File: pytest.ini (hunk 5 above)
Comment (harupy) [reply pytest.ini]: We recently dropped support for python 3.6 and started using python 3.7 in CI, which allows us to use NumPy >=1.20.0 that raises warnings for deprecated type aliases.

File: mlflow/types/utils.py [hunk 8]
```diff
@@ -238,7 +237,7 @@ def __call__(self, x):
...
                 "be mapped iff all values have identical data type which is one "
```
Comment (BenWilson2) [inline mlflow/types/utils.py:241]:
```suggestion
                "be mapped if all values have identical data type which is one "
```

File: mlflow/types/utils.py (hunk 6 above)
Comment (BenWilson2) [reply mlflow/types/utils.py:240]: I think it sounds ok and is similar to how this sort of Exception message would be raised in other libraries.

**Review by BenWilson2 — APPROVED** [review]
LGTM. Nice redundant test removal too!

File: mlflow/types/utils.py (hunk 8 above)
Comment (harupy) [reply mlflow/types/utils.py:241]: `iff` could be intentional: https://github.com/apache/arrow/pull/12221

Comment (BenWilson2) [reply mlflow/types/utils.py:241]: Ah yes... logical operator shorthand (sorry, it's been a long day). Do we want to have something like that in a DS-focused Exception? (I'm just curious now)

File: pytest.ini [hunk 9]
```diff
@@ -1,2 +1,6 @@
 [pytest]
 addopts = --color=yes --durations=5 --showlocals --verbose
+# Prevent deprecated numpy type aliases from being used
+filterwarnings =
+  error:^`np\.[a-z]+` is a deprecated alias for.+:DeprecationWarning:mlflow
+  error:^`np\.[a-z]+` is a deprecated alias for.+:DeprecationWarning:tests
```
Comment (harupy) [inline pytest.ini:6]: This prevents deprecated numpy type aliases from being used

File: mlflow/pyspark/ml/__init__.py [hunk 10]
```diff
@@ -530,7 +530,7 @@ def is_metric_value_loggable(metric_value):
         as an MLflow metric.
         """
         return isinstance(metric_value, (int, float, np.number)) and not isinstance(
-            metric_value, bool
+            metric_value, (bool, np.bool_)
```
Comment (WeichenXu123) [inline mlflow/pyspark/ml/__init__.py:533]: Add `np.bool_` check.

Comment (WeichenXu123) [reply mlflow/pyspark/ml/__init__.py:533]: My original intention is to filter out bool values (including python bool type and numpy bool type). But original code I wrote `np.bool` but actually it should be `np.bool_` according to my original aim.

Comment (WeichenXu123) [reply mlflow/pyspark/ml/__init__.py:533]: Create separate PR https://github.com/mlflow/mlflow/pull/5462

**Review by tomasatdatabricks — APPROVED** [review]
lgtm!

**Pull request merged**